src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -836,6 +836,52 @@ void C2_MacroAssembler::fast_unlock_lightweight(Register obj, Register reg_rax,
   // C2 uses the value of ZF to determine the continuation.
 }
 
+#ifdef ASSERT
+void C2_MacroAssembler::checked_cast_int(const TypeInt* type, Register dst) {
+  BLOCK_COMMENT("CastII {");
+  Label fail;
+  Label succeed;
+  cmpl(dst, type->_lo);
+  jccb(Assembler::less, fail);
+  cmpl(dst, type->_hi);
+  jccb(Assembler::lessEqual, succeed);
+  bind(fail);
+  movl(rax, dst);
+  movl(rcx, type->_lo);
+  movl(rdx, type->_hi);
+  hlt(); // hlt so we have the stack trace
```

Comment:
That's interesting. Sounds like a problem in `NativeStackPrinter::print_stack()`.

Speaking of debugging output, a call into a local helper function (encapsulating pretty printing logic) followed by a hlt call will do the job. But, considering the usages are in-line and quite common I suggest to make it conditional (guarded by a flag). It is possible to recover all 3 values from generated code if needed and turn on error reporting (specify the diagnostic flag) when reproducing failures.

Reply:
I have made a helper function that will print the error message together with the parameters, the stack printing is still problematic, though.